src/realm/collection_parent.hpp

```diff
@@ -73,54 +71,62 @@ enum class UpdateStatus {
 
 /*
  * In order to detect stale collection objects (objects referring to entities that have
- * been deleted from the DB) nested directly in an Obj object, we need a structure that
- * both holds an index of the relevant column as well as a somewhat unique key. The key
- * is generated when the collection is assigned to the property and stored alongside the
- * ref of the collection. The stored key is regenerated/cleared when a new value is
- * assigned to the property.
+ * been deleted from the DB), we need a structure that both holds a somewhat unique salt
+ * and possibly an index of the relevant column. The salt is generated when the collection
+ * is assigned to the property and stored alongside the ref of the collection. The stored
+ * salt is regenerated/cleared when a new value is assigned to the property/collection
+ * element.
  */
-class ColIndex {
+class StableIndex {
 public:
-    ColIndex()
+    StableIndex()
     {
-        value.col_index = 0x7fff;
+        value.raw = 0;
     }
-    ColIndex(ColKey col_key, int64_t key)
+    StableIndex(ColKey col_key, int64_t salt)
     {
         value.col_index = col_key.get_index().val;
         value.is_collection = col_key.is_collection();
-        value.key = uint16_t(key);
+        value.is_column = true;
+        value.salt = int32_t(salt);
     }
-    ColKey::Idx get_index() const noexcept
+    StableIndex(int64_t salt)
     {
-        return {unsigned(value.col_index)};
+        value.raw = 0;
+        value.salt = int32_t(salt);
     }
-    int64_t get_key() const noexcept
+    int64_t get_salt() const
     {
-        return int16_t(value.key);
+        return value.salt;
+    }
+    ColKey::Idx get_index() const noexcept
+    {
+        return {unsigned(value.col_index)};
     }
     bool is_collection() const noexcept
     {
         return value.is_collection;
     }
 
-    bool operator==(const ColIndex& other) const noexcept
+    bool operator==(const StableIndex& other) const noexcept
     {
-        // Compare only index
-        return value.col_index == other.value.col_index;
+        return value.is_column ? value.col_index == other.value.col_index : value.salt == other.value.salt;
     }
 
 private:
-    struct {
-        uint32_t col_index : 15;
-        uint32_t is_collection : 1;
-        uint32_t key : 16;
+    union {
+        struct {
+            bool is_column;
+            bool is_collection;
+            int16_t col_index;
+            int32_t salt;
+        };
+        int64_t raw;
```

Comment:
What's the purpose of `raw`?  Is It for a mixed column that does not contain a collection in order to pass this check `static_assert(sizeof(StableIndex) == 8);`?

Reply:
OK, that's fine. I would have gone with a simple constructor on the struct.